Library/Homebrew/cmd/upgrade.rb

```diff
@@ -171,4 +181,145 @@ def upgrade_formula(f)
+        end
+      end
+
+      u, p = upgradable_dependents(kegs, dependents_to_upgrade)
+
+      formulae_to_upgrade.merge u
+      formulae_pinned.merge p
+    end
+
+    [formulae_to_upgrade, formulae_pinned]
+  end
+
+  def broken_dependents(kegs, formulae)
+    formulae_to_reinstall = {}
+
+    CacheStoreDatabase.use(:linkage) do |db|
+      formulae.each do |formula|
+        dependents_to_reinstall = []
+
+        dependents = kegs.select do |f|
+          f.runtime_dependencies
+           .any? { |d| d["full_name"] == formula.full_name }
+        end
+
+        next if dependents.empty?
+
+        dependents.each do |keg|
+          f = Formulary.from_keg(keg)
+          next if formulae_to_reinstall[f.full_name]
+
+          checker = LinkageChecker.new(keg, cache_db: db)
+          next unless checker.broken_library_linkage?
+
+          dependents_to_reinstall << f
+        end
+
+        r = broken_dependents(kegs, dependents_to_reinstall)
+
+        formulae_to_reinstall.merge r
+      end
+    end
+
+    formulae_to_reinstall
+  end
+
+  def check_dependents(formulae)
+    return if formulae.empty?
+
+    kegs = Formula.installed
+                  .map(&:opt_or_installed_prefix_keg)
+                  .reject { |f| f.runtime_dependencies.to_a.empty? }
+
+    return if kegs.empty?
+
+    # First find all the outdated dependents.
+    oh1 "Checking dependents for outdated formulae" if ARGV.verbose?
+    upgradable, pinned = upgradable_dependents(kegs, formulae)
+
+    # Print the pinned dependents.
+    if !pinned.empty? && !ARGV.include?("--ignore-pinned")
+      ohai "Not upgrading #{Formatter.pluralize(pinned.length, "pinned dependent")}:"
+      puts pinned.values.map { |f| "#{f.full_specified_name} #{f.pkg_version}" } * ", "
+    end
+
+    # Print the upgradable dependents.
+    if upgradable.empty?
+      ohai "No dependents to upgrade" if ARGV.verbose?
+    else
+      ohai "Upgrading #{Formatter.pluralize(upgradable.length, "dependent")}:"
+      formulae_upgrades = upgradable.values.map do |f|
+        if f.optlinked?
+          "#{f.full_specified_name} #{Keg.new(f.opt_prefix).version} -> #{f.pkg_version}"
+        else
+          "#{f.full_specified_name} #{f.pkg_version}"
+        end
+      end
+      puts formulae_upgrades.join(", ")
+    end
+
+    upgrade_formulae(upgradable.values)
+
+    # Assess the dependents tree again.
+
+    kegs = Formula.installed
+                  .map(&:opt_or_installed_prefix_keg)
+                  .reject { |f| f.runtime_dependencies.to_a.empty? }
+
+    oh1 "Checking dependents for broken library links" if ARGV.verbose?
+    reinstallable = broken_dependents(kegs, formulae)
+
+    # Print the broken dependents.
+    if reinstallable.empty?
+      ohai "No broken dependents" if ARGV.verbose?
+    else
+      ohai "Reinstalling #{Formatter.pluralize(reinstallable.length, "broken dependent")} from source:"
+      puts reinstallable.values.map(&:full_specified_name).join(", ")
```

Comment:
Just a thought as I'm not sure if this edge case is handled: if something is pinned and outdated and has broken linkage: if you try to `reinstall` it'll try to install the current version, I think. That case may warrant a special message about it being broken but because it's pinned we can't upgrade it.

Reply:
That's exactly the intention, reinstalling it since we cannot upgrade it. I'll restore the old messaging.

Reply:
@amyspark Just to make sure I'm being clear (I'm not sure I am 😆) I'm proposing that we don't try to `reinstall` pinned, outdated packages with broken linkage but just display an `Error` instead for them.

Reply:
You were not, I apologize 😄 I'll add the partition logic again and print the corresponding error messages.